src/components/ActionButton/StatefulActionButton.ts

```diff
@@ -0,0 +1,131 @@
+import { IResultsComponentBindings } from 'coveo-search-ui';
+import { ActionButton, IActionButtonOptions } from './ActionButton';
+
+/**
+ * Represent a state that can be used by a StatefulActionButton.
+ */
+export interface IStatefulActionButtonState extends IActionButtonOptions {
+    /**
+     * Called when this state is set as the current state.
+     * Called after the onStateExit of the previous state if any.
+     */
+    onStateEntry?: (this: StatefulActionButton) => void;
+    /**
+     * Called when this state has been set as the current state and another is about to get set.
+     * Called before the onStateEntry of the next state if any.
+     */
+    onStateExit?: (this: StatefulActionButton) => void;
+}
+
+/**
+ * Represent a transition from one `IStatefulActionButtonState` to another.
+ */
+export interface IStatefulActionButtonTransition {
+    from: IStatefulActionButtonState;
+    to: IStatefulActionButtonState;
+}
+
+/**
+ * The options format for a StatefulActionButton
+ */
+export interface IStatefulActionButtonOptions {
+    /**
+     * An array containing all states used by the StatefulActionButton instance.
+     * If `switchTo` is called with an unknown state, a warning will be emited
+     * and the transition will not occur.
+     */
+    states: IStatefulActionButtonState[];
+    /**
+     * An array containing all the state transitions allowed on the StatefulActionButton instance.
+     * If `switchTo` is called with an illegal transitions , a warning will be emited
+     * and the transition will not occur.
+     */
+    allowedTransitions?: IStatefulActionButtonTransition[];
+    /**
+     * The initial state to be used by the StatefulActionButton instance.
+     * If nullish or missing from the states array, a warning will be emited,
+     * no ActionButton will be constructed and the element will be hidden
+     */
+    initalState: IStatefulActionButtonState;
+}
+
+/**
+ * An action button able to handle multiple states and their transitions.
+ */
+export class StatefulActionButton {
+    static ID = 'StatefulActionButton';
+    private currentState: IStatefulActionButtonState;
+    private innerActionButton: ActionButton;
+
+    constructor(public element: HTMLElement, public options: IStatefulActionButtonOptions, public bindings?: IResultsComponentBindings) {
+        if (!this.areOptionsValid()) {
+            return;
+        }
+        this.currentState = this.options.initalState;
+        this.innerActionButton = new ActionButton(element, { ...this.options.initalState, click: this.handleClick.bind(this) }, bindings);
+    }
+
+    /**
+     * Switch the state of the instance if the state and the transition between the current and new state are allowed.
+     * @param state a state to try to switch to
+     */
+    public switchTo(state: IStatefulActionButtonState) {
+        if (!this.isTransitionAllowed(state)) {
+            console.warn('This state is not allowed on this StatefulActionButton.');
```

Comment:
If would be nice if `state` had a `name` or `key` attribute we could use to format the warning message. It would be easier to troubleshoot.

Reply:
A name could be nice, I agree, if it's only used for logging purposes (i.e. it's just a name, not a stateId ;) )

Reply:
Yes, that's what I had in mind (using an informative name). Indeed, we could not use it to compare states.